### PR TITLE
Adding sharding CLI params to helper

### PR DESCRIPTION
**Pull request by cberkhoff**
This PR adds sharding parameters to the `helper` binary to start the HTTP shard server (alongside the MPC server). The existing `network.toml` argument is expected to describe the entire network, with sharding peers included. Each pod in this configuration will have 2 ports, one for MPC and one for sharding. The certificate will be shared among those two. The `network.toml` file is expected to have a flat structure with all peer configs as a simple list. 

Here's an example with certificate removed for clarity. Note how we added a new field `shard_port`, that's required for sharding configurations.

```
[[peers]]
certificate = "..."
url = "helper1.prod.ipa-helper.shard0.dev:443"
shard_url = "helper1.prod.ipa-helper.shard0.dev:444"

[peers.hpke]
public_key = "f458d5e1989b2b8f5dacd4143276aa81eaacf7449744ab1251ff667c43550756"

[[peers]]
certificate = "..."
url = "helper2.prod.ipa-helper.shard0.dev:443"
shard_url = "helper2.prod.ipa-helper.shard0.dev:444"

[peers.hpke]
public_key = "62357179868e5594372b801ddf282c8523806a868a2bff2685f66aa05ffd6c22"

[[peers]]
certificate = "..."
url = "helper3.prod.ipa-helper.shard0.dev:443"
url = "helper3.prod.ipa-helper.shard0.dev:444"

[peers.hpke]
public_key = "55f87a8794b4de9a60f8ede9ed000f5f10c028e22390922efc4fb63bc6be0a61"

[[peers]]
certificate = "..."
url = "helper1.prod.ipa-helper.shard1.dev:443"
url = "helper1.prod.ipa-helper.shard1.dev:444"

[peers.hpke]
public_key = "f458d5e1989b2b8f5dacd4143276aa81eaacf7449744ab1251ff667c43550756"

[[peers]]
certificate = "..."
url = "helper2.prod.ipa-helper.shard1.dev:443"
url = "helper2.prod.ipa-helper.shard1.dev:444"

[peers.hpke]
public_key = "62357179868e5594372b801ddf282c8523806a868a2bff2685f66aa05ffd6c22"

[[peers]]
certificate = "..."
url = "helper3.prod.ipa-helper.shard1.dev:443"
url = "helper3.prod.ipa-helper.shard1.dev:444"

[peers.hpke]
public_key = "55f87a8794b4de9a60f8ede9ed000f5f10c028e22390922efc4fb63bc6be0a61"
...
```

### CODE REVIEW

**Review by akoshelev — COMMENTED** [review]
overall I would prefer multiple of 3 configuration, but if this gets us to the goal faster, I don't strongly oppose this design

File: ipa-core/tests/helper_networks.rs
Comment (cberkhoff) [reply ipa-core/tests/helper_networks.rs:74]: I didn't wanted to change this right away but as part of the e2e sharding tests

File: ipa-core/src/config.rs
Comment (cberkhoff) [reply ipa-core/src/config.rs:130]: I think these file conventions are important since they are quite sticky. I will create a separate struct just for that serialization and keep NetworkConfig as is. Let me know what you think

File: ipa-core/tests/helper_networks.rs
Comment (akoshelev) [reply ipa-core/tests/helper_networks.rs:74]: can you please leave a comment here that explains what needs to be changed? I guess I'll run into it soon, but it is good to be explicit

Comment (cberkhoff) [reply ipa-core/tests/helper_networks.rs:74]: This PR is about adding new parameter to `helper` to start both sharded and mpc http servers. I wanted this to help unblock Richa to test invoking this from k8s. Now that `helper` needs sharding parameters, in particular ports on which to run, those need to be provided by the tests.

File: ipa-core/src/config.rs
Comment (akoshelev) [inline ipa-core/src/config.rs:1017]: The inconsistency between ways to specify shard and mpc ports is worrying but not enough to block this change.

File: ipa-core/src/bin/helper.rs
Comment (akoshelev) [inline ipa-core/src/bin/helper.rs:75]:
```suggestion
    /// Port to use for shard-to-shard communication, if sharded MPC is used
    #[arg(default_value = "6000")]
    shard_port: Option<u16>,
```

**Review by akoshelev — APPROVED** [review]
mostly lgtm aside from few minor things

File: ipa-core/src/config.rs
Comment (cberkhoff) [reply ipa-core/src/config.rs:1017]: I'll use sharded_url then

File: ipa-core/src/config.rs
Comment (akoshelev) [inline ipa-core/src/config.rs:1062]: love the port - DOOM was my favourite game when I was a teenager

**Review by akoshelev — APPROVED** [review]

File: ipa-core/src/config.rs
Comment (cberkhoff) [reply ipa-core/src/config.rs:1062]: lol 😈

**Comment by andyleiserson** [comment]
Possibly some updates to draft are needed to support this PR? You can see the error I am getting at https://draft-mpc.vercel.app/query/view/chief-seam2024-11-15T2004. Unfortunately the error doesn't directly indicate what the problem is. (My speculation is that there is an early error return in `helper::server` which leads to the runtime being dropped directly in async code, before reaching the call to `shutdown_background`, resulting in the panic.)

**Comment by cberkhoff** [comment]
This is the error I found:
```
2024-11-15T20:05:27.422429 - 2024-11-15T20:05:27.418438Z ERROR ipa_core::error: ThreadId(1) "main" panicked at /home/ec2-user/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.37.0/src/runtime/blocking/shutdown.rs:51:21:
2024-11-15T20:05:27.423215 - Cannot drop a runtime in a context where blocking is not allowed. This happens when a runtime is dropped from within an asynchronous context.
2024-11-15T20:05:27.423552 - stack trace:
2024-11-15T20:05:27.423830 -    0: ipa_core::error::set_global_panic_hook::{{closure}}
2024-11-15T20:05:27.424064 -    1: std::panicking::rust_panic_with_hook
2024-11-15T20:05:27.424289 -    2: std::panicking::begin_panic_handler::{{closure}}
2024-11-15T20:05:27.424508 -    3: std::sys::backtrace::__rust_end_short_backtrace
2024-11-15T20:05:27.424722 -    4: rust_begin_unwind
2024-11-15T20:05:27.424933 -    5: core::panicking::panic_fmt
2024-11-15T20:05:27.425153 -    6: tokio::runtime::blocking::shutdown::Receiver::wait
2024-11-15T20:05:27.425370 -    7: tokio::runtime::blocking::pool::BlockingPool::shutdown
2024-11-15T20:05:27.425583 -    8: core::ptr::drop_in_place<tokio::runtime::blocking::pool::BlockingPool>
2024-11-15T20:05:27.425794 -    9: helper::server::{{closure}}
2024-11-15T20:05:27.426004 -   10: <core::pin::Pin<P> as core::future::future::Future>::poll
2024-11-15T20:05:27.426212 -   11: tokio::runtime::scheduler::current_thread::Context::enter
2024-11-15T20:05:27.426418 -   12: tokio::runtime::context::scoped::Scoped<T>::set
2024-11-15T20:05:27.426627 -   13: tokio::runtime::scheduler::current_thread::CoreGuard::block_on
2024-11-15T20:05:27.426835 -   14: tokio::runtime::context::runtime::enter_runtime
2024-11-15T20:05:27.427141 -   15: tokio::runtime::runtime::Runtime::block_on
2024-11-15T20:05:27.427356 -   16: helper::main
2024-11-15T20:05:27.427566 -   17: std::sys::backtrace::__rust_begin_short_backtrace
2024-11-15T20:05:27.427786 -   18: std::rt::lang_start::{{closure}}
2024-11-15T20:05:27.428003 -   19: std::rt::lang_start_internal
2024-11-15T20:05:27.428215 -   20: main
2024-11-15T20:05:27.428428 -   21: __libc_start_call_main
2024-11-15T20:05:27.428637 -   22: __libc_start_main_alias_1
2024-11-15T20:05:27.428846 -   23: _start
2024-11-15T20:05:27.429055 - 
2024-11-15T20:05:27.429263 - thread 'main' panicked at /home/ec2-user/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.37.0/src/runtime/blocking/shutdown.rs:51:21:
2024-11-15T20:05:27.429508 - Cannot drop a runtime in a context where blocking is not allowed. This happens when a runtime is dropped from within an asynchronous context.
2024-11-15T20:05:27.429723 - note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
2024-11-15T20:05:27.430621 - updating status: status=<Status.CRASHED: 9>
```

**Comment by akoshelev** [comment]
as Andy pointed, it does not seem that this is the root cause - runtime is likely getting dropped because of a panic in `2024-11-15T20:05:27.425794 -    9: helper::server::{{closure}}`. In release builds it is hard to reproduce, so it is probably required to just run it locally in debug mode

**Pull request merged**
